components/script_plugins/unrooted_must_root.rs

```diff
@@ -212,4 +210,32 @@ impl<'a, 'b, 'tcx> MirVisitor<'tcx> for MirFnVisitor<'a, 'b, 'tcx> {
             _ => {},
         }
     }
+
+    fn visit_constant(&mut self, constant: &mir::Constant<'tcx>, location: mir::Location) {
+        self.super_constant(constant, location);
+
+        let ur_cx = &self.unrooted_cx;
+        let cx = ur_cx.late_cx;
+        match constant.ty.sty {
+            ty::TyFnDef(callee_def_id, callee_substs) => {
+                let callee_generics = cx.tcx.generics_of(callee_def_id);
+                for ty_param_def in &callee_generics.types {
+                    // If type has `#[must_root]`, then it is ok to
+                    // give it a must-root type, so just skip.
+                    if cx.tcx.has_attr(ty_param_def.def_id, "must_root") {
+                        continue;
+                    }
+
+                    let arg_ty = callee_substs.type_at(ty_param_def.index as usize);
+                    if ur_cx.is_unrooted_ty(arg_ty, false) {
+                        cx.span_lint(UNROOTED_MUST_ROOT, constant.span, "Callee generic type must be rooted.")
+                    }
+                }
+            }
+            ty::TyAdt(_, _) => {
+                cx.span_lint(UNROOTED_MUST_ROOT, constant.span, "xd")
```

Comment:
you probably don't need this here -- this corresponds to a constant value of ADT type, e.g. `const FOO: Option<i32> = Some(22)` or something. This arm is checking the *final type* of the constant, and I think that's better done elsewhere. We only wanted to catch the case where we are supply unsuitable values for `#[must_root]` type parameters.

Reply:
Oh, I forgot to remove this check. I was just playing around and then cleaning the code late in the night, so I missed it. Thanks for the explanation what it actually does!